internal/expr/parser.go

```diff
@@ -17,17 +18,33 @@ type queryPart interface {
 	part()
 }
 
-// valueAccessor stores information for accessing a Go value. It consists of a
-// type name and some value within it to be accessed. For example: a field of a
-// struct, or a key of a map.
-type valueAccessor struct {
+// valueAccessor stores information about how to access a Go value. For example,
+// by index, by key, or by slice syntax.
+type valueAccessor interface {
+	TypeName() string
+
+	// marker method
+	accessor()
+}
+
+// keyedAccesor stores information for accessing a keyed Go value. It consists
+// of a type name and some value within it to be accessed. For example: a field
+// of a struct, or a key of a map.
+type keyedAccesor struct {
 	typeName, memberName string
 }
 
-func (va valueAccessor) String() string {
```

Comment:
The problem with using a pointer is that I also have to allocate the struct on the heap if we are returning a `valueAccessor` from a function, in this case from `parseSliceAccessor`. And, I would have to check that the returned pointer is not null before using it. Do you still think pointer receiver is better even with theses caveats?

Reply:
I think it would be fine not to be checking for nil for this value, we would only return a nil pointer to one in an error case. As for assigning on the heap, I'm not sure what would happen exactly in the case/version of the compiler, but interface variables with an interface type go on the heap according to this https://research.swtch.com/interfaces.

But I think its fair enough to not change this because of the extra complexity that it would add, especially since it contains a very small amount of information.